File: Introduction_to_numerical_programming_using_Python_and_CPP_Beu/Ch04/C/P04-Sort1.cpp
// Ascending sort and indexing of an array
#include <stdio.h>
#include "sort.h"

int main()
{
   const int n = 6;                          // number of values to be sorted
   double x[] = {0., 30., 60., 50., 20., 10., 40.};     // array to be sorted
   int ind[n+1], rnk[n+1];                     // arrays of indexes and ranks
   int i;

   printf("Original array:\n");
   for (i=1; i<=n; i++) printf("%6.2f",x[i]); printf("\n");

   Index(x,ind,n);
   printf("Indexes:\n");
   for (i=1; i<=n; i++) printf("%6i",ind[i]); printf("\n");

   Rank(ind,rnk,n);
   printf("Ranks:\n");
   for (i=1; i<=n; i++) printf("%6i",rnk[i]); printf("\n");

   printf("Indexed array:\n");
   for (i=1; i<=n; i++) printf("%6.2f",x[ind[i]]); printf("\n");

   InsertSort(x,n);
   printf("Sorted array:\n");
   for (i=1; i<=n; i++) printf("%6.2f",x[i]); printf("\n");
}

File: Introduction_to_numerical_programming_using_Python_and_CPP_Beu/Ch04/Python/P04-Sort1.py
# Ascending sort and indexing of an array
from sort import *

# main

n = 6                                         # number of values to be sorted
x = [0., 30., 60., 50., 20., 10., 40.]                   # array to be sorted
ind = [0]*(n+1); rnk = [0]*(n+1)                # arrays of indexes and ranks

print("Original array:")
for i in range(1,n+1): print("{0:6.2f}".format(x[i]),end="")
print()

Index(x,ind,n)
print("Indexes:")
for i in range(1,n+1): print("{0:6d}".format(ind[i]),end="")
print()

Rank(ind,rnk,n)
print("Ranks:")
for i in range(1,n+1): print("{0:6d}".format(rnk[i]),end="")
print()

print("Indexed array:")
for i in range(1,n+1): print("{0:6.2f}".format(x[ind[i]]),end="")
print()

InsertSort(x,n)
print("Sorted array:")
for i in range(1,n+1): print("{0:6.2f}".format(x[i]),end="")
print()
